src/xdp/ebpfextension.c

```diff
@@ -0,0 +1,277 @@
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+//
+
+#include "precomp.h"
+#include "ebpfextension.tmh"
+
+typedef struct _EBPF_EXTENSION_PROVIDER {
+    NPI_PROVIDER_CHARACTERISTICS Characteristics;
+    HANDLE NmrProviderHandle;
+    EBPF_EXTENSION_ON_CLIENT_ATTACH *AttachCallback;
+    EBPF_EXTENSION_ON_CLIENT_DETACH *DetachCallback;
+    const VOID *CustomData;
+} EBPF_EXTENSION_PROVIDER;
+
+typedef struct _EBPF_EXTENSION_CLIENT {
+    HANDLE NmrBindingHandle;
+    GUID ClientModuleId;
+    const ebpf_extension_dispatch_table_t *ClientDispatch;
+    const VOID *ClientBindingContext;
+    const ebpf_extension_data_t *ClientData;
+    VOID *ProviderData;
+    EBPF_EXTENSION_PROVIDER *ProviderContext;
+} EBPF_EXTENSION_CLIENT;
+
+VOID
+EbpfExtensionClientSetProviderData(
+    _In_ const EBPF_EXTENSION_CLIENT *Client,
+    _In_opt_ const VOID *ProviderData
+    )
+{
+    ((EBPF_EXTENSION_CLIENT *)Client)->ProviderData = (VOID *)ProviderData;
+}
+
+VOID *
+EbpfExtensionClientGetProviderData(
+    _In_ const EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    return Client->ProviderData;
+}
+
+const ebpf_extension_dispatch_table_t *
+EbpfExtensionClientGetDispatch(
+    _In_ const EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    return Client->ClientDispatch;
+}
+
+const VOID *
+EbpfExtensionClientGetClientContext(
+    _In_ const EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    return Client->ClientBindingContext;
+}
+
+const ebpf_extension_data_t *
+EbpfExtensionClientGetClientData(
+    _In_ const EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    return Client->ClientData;
+}
+
+VOID
+EbpfExtensionDetachClientCompletion(
+    _In_ EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    TraceEnter(TRACE_CORE, "Client=%p", Client);
+
+    ASSERT(Client != NULL);
+    _Analysis_assume_(Client != NULL);
+
+    NmrProviderDetachClientComplete(Client->NmrBindingHandle);
+
+    TraceExitSuccess(TRACE_CORE);
+}
+
+static
+NTSTATUS
+EbpfExtensionProviderAttachClient(
+    _In_ HANDLE NmrBindingHandle,
+    _In_ const VOID *ProviderContext,
+    _In_ const NPI_REGISTRATION_INSTANCE* ClientRegistrationInstance,
+    _In_ const VOID *ClientBindingContext,
+    _In_ const VOID *ClientNpiDispatch,
+    _Outptr_ VOID **ProviderBindingContext,
+    _Outptr_result_maybenull_ const VOID **ProviderDispatch
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    EBPF_EXTENSION_PROVIDER *Provider = (EBPF_EXTENSION_PROVIDER *)ProviderContext;
+    EBPF_EXTENSION_CLIENT *Client = NULL;
+    const ebpf_extension_dispatch_table_t *ClientDispatch;
+
+    TraceEnter(TRACE_CORE, "ProviderContext=%p", ProviderContext);
+
+    if ((ProviderBindingContext == NULL) || (ProviderDispatch == NULL) || (Provider == NULL)) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    *ProviderBindingContext = NULL;
+    *ProviderDispatch = NULL;
```

Comment:
Since `NTSTATUS` is already defined as `typedef _Return_type_success_(return >= 0) LONG NTSTATUS;` that means this function has been explicitly annotated to only promise to return these two if you return a success code; so this isn't strictly necessary. It's definitely a NIT, but I personally prefer not to do this. I leave it up to you though.

Reply:
Sure, I refactored, which also clarified the actual flow through the function.